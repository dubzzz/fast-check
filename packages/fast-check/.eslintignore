bundle/
node_modules/
lib/
lib-*/
perf/
jest*.js
jest*.cjs