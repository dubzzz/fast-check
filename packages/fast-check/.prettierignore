bundle/
coverage/
docs/
documentation/Arbitraries.md
lib/
lib-*/