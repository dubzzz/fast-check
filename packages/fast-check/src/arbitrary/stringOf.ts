import type { Arbitrary } from '../check/arbitrary/definition/Arbitrary';
import type { ArrayConstraintsInternal } from './array';
import { array } from './array';
import type { StringSharedConstraints } from './_shared/StringSharedConstraints';
import { patternsToStringMapper, patternsToStringUnmapperFor } from './_internals/mappers/PatternsToString';
import { createSlicesForString } from './_internals/helpers/SlicesForStringBuilder';
export type { StringSharedConstraints } from './_shared/StringSharedConstraints';

const safeObjectAssign = Object.assign;

/**
 * For strings using the characters produced by `charArb`
 *
 * @param charArb - Arbitrary able to generate random strings (possibly multiple characters)
 * @param constraints - Constraints to apply when building instances (since 2.4.0)
 *
 * @remarks Since 1.1.3
 * @public
 */
export function stringOf(charArb: Arbitrary<string>, constraints: StringSharedConstraints = {}): Arbitrary<string> {
  const unmapper = patternsToStringUnmapperFor(charArb, constraints);
  const experimentalCustomSlices = createSlicesForString(charArb, unmapper);
  // TODO - Move back to object spreading as soon as we bump support from es2017 to es2018+
  const enrichedConstraints: ArrayConstraintsInternal<string> = safeObjectAssign(safeObjectAssign({}, constraints), {
    experimentalCustomSlices,
  });
  return array(charArb, enrichedConstraints).map(patternsToStringMapper, unmapper);
}
