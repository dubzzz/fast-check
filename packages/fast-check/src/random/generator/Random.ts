import type { RandomGenerator } from 'pure-rand';
import {
  unsafeUniformArrayIntDistribution,
  unsafeUniformBigIntDistribution,
  unsafeUniformIntDistribution,
} from 'pure-rand';

/**
 * Wrapper around an instance of a `pure-rand`'s random number generator
 * offering a simpler interface to deal with random with impure patterns
 *
 * @public
 */
export class Random {
  private static MIN_INT: number = 0x80000000 | 0;
  private static MAX_INT: number = 0x7fffffff | 0;
  private static DBL_FACTOR: number = Math.pow(2, 27);
  private static DBL_DIVISOR: number = Math.pow(2, -53);

  private internalRng: RandomGenerator;

  /**
   * Create a mutable random number generator by cloning the passed one and mutate it
   * @param sourceRng - Immutable random generator from pure-rand library, will not be altered (a clone will be)
   */
  constructor(sourceRng: RandomGenerator) {
    this.internalRng = sourceRng.clone();
  }

  /**
   * Clone the random number generator
   */
  clone(): Random {
    return new Random(this.internalRng);
  }

  /**
   * Generate an integer having `bits` random bits
   * @param bits - Number of bits to generate
   */
  next(bits: number): number {
    return unsafeUniformIntDistribution(0, (1 << bits) - 1, this.internalRng);
  }

  /**
   * Generate a random boolean
   */

  nextBoolean(): boolean {
    return unsafeUniformIntDistribution(0, 1, this.internalRng) == 1;
  }

  /**
   * Generate a random integer (32 bits)
   */
  nextInt(): number;

  /**
   * Generate a random integer between min (included) and max (included)
   * @param min - Minimal integer value
   * @param max - Maximal integer value
   */
  nextInt(min: number, max: number): number;
  nextInt(min?: number, max?: number): number {
    return unsafeUniformIntDistribution(
      min == null ? Random.MIN_INT : min,
      max == null ? Random.MAX_INT : max,
      this.internalRng,
    );
  }

  /**
   * Generate a random bigint between min (included) and max (included)
   * @param min - Minimal bigint value
   * @param max - Maximal bigint value
   */
  nextBigInt(min: bigint, max: bigint): bigint {
    return unsafeUniformBigIntDistribution(min, max, this.internalRng);
  }

  /**
   * Generate a random ArrayInt between min (included) and max (included)
   * @param min - Minimal ArrayInt value
   * @param max - Maximal ArrayInt value
   */
  nextArrayInt(
    min: { sign: 1 | -1; data: number[] },
    max: { sign: 1 | -1; data: number[] },
  ): { sign: 1 | -1; data: number[] } {
    return unsafeUniformArrayIntDistribution(min, max, this.internalRng);
  }

  /**
   * Generate a random floating point number between 0.0 (included) and 1.0 (excluded)
   */
  nextDouble(): number {
    const a = this.next(26);
    const b = this.next(27);
    return (a * Random.DBL_FACTOR + b) * Random.DBL_DIVISOR;
  }
}
