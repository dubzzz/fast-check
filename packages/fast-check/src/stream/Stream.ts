import {
  filterHelper,
  flatMapHelper,
  joinHelper,
  mapHelper,
  nilHelper,
  takeNHelper,
  takeWhileHelper,
} from './StreamHelpers';

const safeSymbolIterator: typeof Symbol.iterator = Symbol.iterator;

/**
 * Wrapper around `IterableIterator` interface
 * offering a set of helpers to deal with iterations in a simple way
 *
 * @remarks Since 0.0.7
 * @public
 */
export class Stream<T> implements IterableIterator<T> {
  /**
   * Create an empty stream of T
   * @remarks Since 0.0.1
   */
  static nil<T>(): Stream<T> {
    return new Stream(nilHelper());
  }

  /**
   * Create a stream of T from a variable number of elements
   *
   * @param elements - Elements used to create the Stream
   * @remarks Since 2.12.0
   */
  static of<T>(...elements: T[]): Stream<T> {
    return new Stream(elements[safeSymbolIterator]());
  }

  // /*DEBUG*/ // no double iteration
  // /*DEBUG*/ private isLive: boolean;

  /**
   * Create a Stream based on `g`
   * @param g - Underlying data of the Stream
   */
  constructor(/** @internal */ private readonly g: IterableIterator<T>) {
    // /*DEBUG*/ this.isLive = true;
  }

  // /*DEBUG*/ private closeCurrentStream() {
  // /*DEBUG*/   if (! this.isLive) throw new Error('Stream has already been closed');
  // /*DEBUG*/   this.isLive = false;
  // /*DEBUG*/ }

  next(): IteratorResult<T> {
    return this.g.next();
  }
  [safeSymbolIterator](): IterableIterator<T> {
    // /*DEBUG*/ this.closeCurrentStream();
    return this.g;
  }

  /**
   * Map all elements of the Stream using `f`
   *
   * WARNING: It closes the current stream
   *
   * @param f - Mapper function
   * @remarks Since 0.0.1
   */
  map<U>(f: (v: T) => U): Stream<U> {
    // /*DEBUG*/ this.closeCurrentStream();
    return new Stream(mapHelper(this.g, f));
  }
  /**
   * Flat map all elements of the Stream using `f`
   *
   * WARNING: It closes the current stream
   *
   * @param f - Mapper function
   * @remarks Since 0.0.1
   */
  flatMap<U>(f: (v: T) => IterableIterator<U>): Stream<U> {
    // /*DEBUG*/ this.closeCurrentStream();
    return new Stream(flatMapHelper(this.g, f));
  }

  /**
   * Drop elements from the Stream while `f(element) === true`
   *
   * WARNING: It closes the current stream
   *
   * @param f - Drop condition
   * @remarks Since 0.0.1
   */
  dropWhile(f: (v: T) => boolean): Stream<T> {
    let foundEligible = false;
    function* helper(v: T): IterableIterator<T> {
      if (foundEligible || !f(v)) {
        foundEligible = true;
        yield v;
      }
    }
    return this.flatMap(helper);
  }
  /**
   * Drop `n` first elements of the Stream
   *
   * WARNING: It closes the current stream
   *
   * @param n - Number of elements to drop
   * @remarks Since 0.0.1
   */
  drop(n: number): Stream<T> {
    if (n <= 0) {
      return this;
    }
    let idx = 0;
    function helper(): boolean {
      return idx++ < n;
    }
    return this.dropWhile(helper);
  }
  /**
   * Take elements from the Stream while `f(element) === true`
   *
   * WARNING: It closes the current stream
   *
   * @param f - Take condition
   * @remarks Since 0.0.1
   */
  takeWhile(f: (v: T) => boolean): Stream<T> {
    // /*DEBUG*/ this.closeCurrentStream();
    return new Stream(takeWhileHelper(this.g, f));
  }
  /**
   * Take `n` first elements of the Stream
   *
   * WARNING: It closes the current stream
   *
   * @param n - Number of elements to take
   * @remarks Since 0.0.1
   */
  take(n: number): Stream<T> {
    // /*DEBUG*/ this.closeCurrentStream();
    return new Stream(takeNHelper(this.g, n));
  }

  /**
   * Filter elements of the Stream
   *
   * WARNING: It closes the current stream
   *
   * @param f - Elements to keep
   * @remarks Since 1.23.0
   */
  filter<U extends T>(f: (v: T) => v is U): Stream<U>;
  /**
   * Filter elements of the Stream
   *
   * WARNING: It closes the current stream
   *
   * @param f - Elements to keep
   * @remarks Since 0.0.1
   */
  filter(f: (v: T) => boolean): Stream<T>;
  filter<U extends T>(f: (v: T) => v is U): Stream<U> {
    // /*DEBUG*/ this.closeCurrentStream();
    return new Stream(filterHelper(this.g, f));
  }

  /**
   * Check whether all elements of the Stream are successful for `f`
   *
   * WARNING: It closes the current stream
   *
   * @param f - Condition to check
   * @remarks Since 0.0.1
   */
  every(f: (v: T) => boolean): boolean {
    // /*DEBUG*/ this.closeCurrentStream();
    for (const v of this.g) {
      if (!f(v)) {
        return false;
      }
    }
    return true;
  }
  /**
   * Check whether one of the elements of the Stream is successful for `f`
   *
   * WARNING: It closes the current stream
   *
   * @param f - Condition to check
   * @remarks Since 0.0.1
   */
  has(f: (v: T) => boolean): [boolean, T | null] {
    // /*DEBUG*/ this.closeCurrentStream();
    for (const v of this.g) {
      if (f(v)) {
        return [true, v];
      }
    }
    return [false, null];
  }

  /**
   * Join `others` Stream to the current Stream
   *
   * WARNING: It closes the current stream and the other ones (as soon as it iterates over them)
   *
   * @param others - Streams to join to the current Stream
   * @remarks Since 0.0.1
   */
  join(...others: IterableIterator<T>[]): Stream<T> {
    // /*DEBUG*/ this.closeCurrentStream();
    return new Stream(joinHelper(this.g, others));
  }

  /**
   * Take the `nth` element of the Stream of the last (if it does not exist)
   *
   * WARNING: It closes the current stream
   *
   * @param nth - Position of the element to extract
   * @remarks Since 0.0.12
   */
  getNthOrLast(nth: number): T | null {
    // /*DEBUG*/ this.closeCurrentStream();
    let remaining = nth;
    let last: T | null = null;
    for (const v of this.g) {
      if (remaining-- === 0) return v;
      last = v;
    }
    return last;
  }
}

/**
 * Create a Stream based on `g`
 *
 * @param g - Underlying data of the Stream
 *
 * @remarks Since 0.0.7
 * @public
 */
export function stream<T>(g: IterableIterator<T>): Stream<T> {
  return new Stream<T>(g);
}
