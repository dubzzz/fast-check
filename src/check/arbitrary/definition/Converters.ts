import { Arbitrary, assertIsArbitrary } from './Arbitrary';
import { ArbitraryWithContextualShrink } from './ArbitraryWithContextualShrink';
import { ConverterFromNext } from './ConverterFromNext';
import { ConverterToNext } from './ConverterToNext';
import { assertIsNextArbitrary, NextArbitrary } from './NextArbitrary';

/**
 * Convert an instance of NextArbitrary to an instance of Arbitrary
 * @param arb - The instance to be converted
 * @public
 */
export function convertFromNext<T>(arb: NextArbitrary<T>): Arbitrary<T> {
  if (ConverterToNext.isConverterToNext(arb)) {
    return arb.arb;
  }
  assertIsNextArbitrary(arb);
  return new ConverterFromNext(arb);
}

/**
 * Convert an instance of NextArbitrary to an instance of ArbitraryWithContextualShrink
 * @param arb - The instance to be converted
 * @param legacyShrunkOnceContext - Default context to be returned when shrunk once
 * @public
 */
export function convertFromNextWithShrunkOnce<T>(
  arb: NextArbitrary<T>,
  legacyShrunkOnceContext: unknown
): ArbitraryWithContextualShrink<T> {
  if (ConverterToNext.isConverterToNext(arb)) {
    if (
      !('contextualShrink' in arb.arb) ||
      !('contextualShrinkableFor' in arb.arb) ||
      !('shrunkOnceContext' in arb.arb) ||
      !('shrink' in arb.arb) ||
      !('shrinkableFor' in arb.arb)
    ) {
      throw new Error('Conversion rejected: Underlying arbitrary is not compatible with ArbitraryWithContextualShrink');
    }
    return arb.arb;
  }
  assertIsNextArbitrary(arb);
  return new ConverterFromNext(arb, legacyShrunkOnceContext);
}

/**
 * Convert an instance of Arbitrary to an instance of NextArbitrary
 * @param arb - The instance to be converted
 * @public
 */
export function convertToNext<T>(arb: Arbitrary<T>): NextArbitrary<T> {
  if (ConverterFromNext.isConverterFromNext(arb)) {
    return arb.arb as NextArbitrary<T>;
  }
  assertIsArbitrary(arb);
  return new ConverterToNext(arb);
}
