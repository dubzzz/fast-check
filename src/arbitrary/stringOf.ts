import { Arbitrary } from '../check/arbitrary/definition/Arbitrary';
import { convertFromNext, convertToNext } from '../check/arbitrary/definition/Converters';
import { array } from './array';
import { StringSharedConstraints } from './_shared/StringSharedConstraints';
import { patternsToStringMapper, patternsToStringUnmapperFor } from './_internals/mappers/PatternsToString';
export { StringSharedConstraints } from './_shared/StringSharedConstraints';

/**
 * For strings using the characters produced by `charArb`
 *
 * @param charArb - Arbitrary able to generate random strings (possibly multiple characters)
 * @param constraints - Constraints to apply when building instances (since 2.4.0)
 *
 * @remarks Since 1.1.3
 * @public
 */
export function stringOf(charArb: Arbitrary<string>, constraints: StringSharedConstraints = {}): Arbitrary<string> {
  return convertFromNext(
    convertToNext(array(charArb, constraints)).map(
      patternsToStringMapper,
      patternsToStringUnmapperFor(convertToNext(charArb), constraints)
    )
  );
}
