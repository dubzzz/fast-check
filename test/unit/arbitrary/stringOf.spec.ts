import * as fc from '../../../lib/fast-check';
import { stringOf } from '../../../src/arbitrary/stringOf';

import { convertFromNext, convertToNext } from '../../../src/check/arbitrary/definition/Converters';
import { NextArbitrary } from '../../../src/check/arbitrary/definition/NextArbitrary';
import { NextValue } from '../../../src/check/arbitrary/definition/NextValue';
import { Random } from '../../../src/random/generator/Random';
import { Stream } from '../../../src/stream/Stream';
import {
  assertGenerateProducesSameValueGivenSameSeed,
  assertGenerateProducesValuesFlaggedAsCanGenerate,
  assertShrinkProducesValuesFlaggedAsCanGenerate,
} from '../check/arbitrary/generic/NextArbitraryAssertions';
import { buildNextShrinkTree, renderTree } from '../check/arbitrary/generic/ShrinkTree';

describe('stringOf (integration)', () => {
  type Extra = { minLength?: number; maxLength?: number; patterns: string[] };
  const extraParameters: fc.Arbitrary<Extra> = fc
    .tuple(
      fc.nat({ max: 30 }),
      fc.nat({ max: 30 }),
      fc.boolean(),
      fc.boolean(),
      // empty string not supported when trying to recognize stuff geenrated by this precise arbitrary
      fc.array(fc.fullUnicodeString({ minLength: 1 }), { minLength: 1 })
    )
    .map(([min, gap, withMin, withMax, patterns]) => ({
      minLength: withMin ? min : undefined,
      maxLength: withMax ? min + gap : undefined,
      patterns,
    }));

  const stringOfBuilder = (extra: Extra) =>
    convertToNext(stringOf(convertFromNext(new PatternsArbitrary(extra.patterns)), extra));

  it('should generate the same values given the same seed', () => {
    assertGenerateProducesSameValueGivenSameSeed(stringOfBuilder, { extraParameters });
  });

  it('should recognize values that would have been generated using it during generate', () => {
    assertGenerateProducesValuesFlaggedAsCanGenerate(stringOfBuilder, { extraParameters });
  });

  it('should shrink towards the same values given the same seed', () => {
    assertGenerateProducesSameValueGivenSameSeed(stringOfBuilder, { extraParameters });
  });

  // assertShrinkProducesSameValueWithoutInitialContext is not fully supported by stringOf
  // Indeed depending how much the source strings overlaps there are possibly many possible ways to build a given string
  // so also many possible ways to shrink it.

  it('should recognize values that would have been generated using it during shrink', () => {
    assertShrinkProducesValuesFlaggedAsCanGenerate(stringOfBuilder, { extraParameters });
  });

  it.each`
    rawValue         | patterns
    ${'CccABbBbCcc'} | ${['A', 'Bb', 'Ccc']}
    ${'_._.____...'} | ${['._', '_...', '_._.', '_..', '.', '.._.', '__.', '....', '..', '.___', '._..', '__', '_.', '___', '.__.', '__._', '._.', '...', '_', '.._', '..._', '.__', '_.._', '_.__', '__..']}
  `('should be able to shrink $rawValue', ({ rawValue, patterns }) => {
    // Arrange
    const arb = convertToNext(stringOf(convertFromNext(new PatternsArbitrary(patterns))));
    const value = new NextValue(rawValue, undefined);

    // Act
    const renderedTree = renderTree(buildNextShrinkTree(arb, value, { numItems: 100 })).join('\n');

    // Assert
    expect(arb.canShrinkWithoutContext(rawValue)).toBe(true);
    expect(renderedTree).toMatchSnapshot();
  });
});

// Helpers

class PatternsArbitrary extends NextArbitrary<string> {
  constructor(readonly patterns: string[]) {
    super();
  }
  generate(mrng: Random): NextValue<string> {
    return new NextValue(this.patterns[mrng.nextInt(0, this.patterns.length - 1)], undefined);
  }
  canShrinkWithoutContext(value: unknown): value is string {
    if (typeof value !== 'string') return false;
    return this.patterns.includes(value);
  }
  shrink(value: string): Stream<NextValue<string>> {
    const patternIndex = this.patterns.indexOf(value);
    if (patternIndex <= 0) {
      return Stream.nil();
    }
    return Stream.of(new NextValue(this.patterns[0], undefined));
  }
}
