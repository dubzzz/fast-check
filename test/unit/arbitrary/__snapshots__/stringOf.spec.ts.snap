// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`stringOf (integration) should be able to shrink _._.____... 1`] = `
"\\"_._.____...\\"
├> \\"\\"
├> \\"___...\\"
|  ├> \\"_...\\"
|  |  ├> \\"...\\"
|  |  |  ├> \\"..\\"
|  |  |  |  ├> \\"\\"
|  |  |  |  └> \\"._\\"
|  |  |  |     └> \\"\\"
|  |  |  ├> \\"._..\\"
|  |  |  |  ├> \\"..\\"
|  |  |  |  |  ├> \\"\\"
|  |  |  |  |  └> \\"._\\"
|  |  |  |  |     └> \\"\\"
|  |  |  |  └> \\"._\\"
|  |  |  |     └> \\"\\"
|  |  |  └> \\".\\"
|  |  |     ├> \\"\\"
|  |  |     └> \\"._\\"
|  |  |        └> \\"\\"
|  |  ├> \\"._...\\"
|  |  |  ├> \\"...\\"
|  |  |  |  ├> \\"..\\"
|  |  |  |  |  ├> \\"\\"
|  |  |  |  |  └> \\"._\\"
|  |  |  |  |     └> \\"\\"
|  |  |  |  ├> \\"._..\\"
|  |  |  |  |  ├> \\"..\\"
|  |  |  |  |  |  ├> \\"\\"
|  |  |  |  |  |  └> \\"._\\"
|  |  |  |  |  |     └> \\"\\"
|  |  |  |  |  └> \\"._\\"
|  |  |  |  |     └> \\"\\"
|  |  |  |  └> \\".\\"
|  |  |  |     ├> \\"\\"
|  |  |  |     └> \\"._\\"
|  |  |  |        └> \\"\\"
|  |  |  ├> \\"._\\"
|  |  |  |  └> \\"\\"
|  |  |  ├> \\"._..\\"
|  |  |  |  ├> \\"..\\"
|  |  |  |  |  ├> \\"\\"
|  |  |  |  |  └> \\"._\\"
|  |  |  |  |     └> \\"\\"
|  |  |  |  └> \\"._\\"
|  |  |  |     └> \\"\\"
|  |  |  └> \\"._.\\"
|  |  |     ├> \\".\\"
|  |  |     |  ├> \\"\\"
|  |  |     |  └> \\"._\\"
|  |  |     |     └> \\"\\"
|  |  |     └> \\"._\\"
|  |  |        └> \\"\\"
|  |  ├> \\"_\\"
|  |  |  ├> \\"\\"
|  |  |  └> \\"._\\"
|  |  |     └> \\"\\"
|  |  ├> \\"_..\\"
|  |  |  ├> \\"..\\"
|  |  |  |  ├> \\"\\"
|  |  |  |  └> \\"._\\"
|  |  |  |     └> \\"\\"
|  |  |  ├> \\"._..\\"
|  |  |  |  ├> \\"..\\"
|  |  |  |  |  ├> \\"\\"
|  |  |  |  |  └> \\"._\\"
|  |  |  |  |     └> \\"\\"
|  |  |  |  └> \\"._\\"
|  |  |  |     └> \\"\\"
|  |  |  └> \\"_\\"
|  |  |     ├> \\"\\"
|  |  |     └> \\"._\\"
|  |  |        └> \\"\\"
|  |  └> \\"_.\\"
|  |     ├> \\".\\"
|  |     |  ├> \\"\\"
|  |     |  └> \\"._\\"
|  |     |     └> \\"\\"
|  |     ├> \\"._.\\"
|  |     |  ├> \\".\\"
|  |     |  |  ├> \\"\\"
|  |     |  |  └> \\"._\\"
|  |     |  |     └> \\"\\"
|  |     |  └> \\"._\\"
|  |     |     └> \\"\\"
|  |     └> \\"_\\"
|  |        ├> \\"\\"
|  |        └> \\"._\\"
|  |           └> \\"\\"
|  ├> \\"__...\\"
|  |  ├> \\"_...\\"
|  |  |  ├> \\"...\\"
|  |  |  |  ├> \\"..\\"
|  |  |  |  |  ├> \\"\\"
|  |  |  |  |  └> \\"._\\"
|  |  |  |  |     └> \\"\\"
|  |  |  |  ├> \\"._..\\"
|  |  |  |  |  ├> \\"..\\"
|  |  |  |  |  |  ├> \\"\\"
|  |  |  |  |  |  └> \\"._\\"
|  |  |  |  |  |     └> …
|  |  |  |  |  └> …
|  |  |  |  └> …
|  |  |  └> …
|  |  └> …
|  └> …
└> …"
`;

exports[`stringOf (integration) should be able to shrink CccABbBbCcc 1`] = `
"\\"CccABbBbCcc\\"
├> \\"\\"
├> \\"BbBbCcc\\"
|  ├> \\"BbCcc\\"
|  |  ├> \\"Ccc\\"
|  |  |  ├> \\"\\"
|  |  |  └> \\"A\\"
|  |  |     └> \\"\\"
|  |  ├> \\"ACcc\\"
|  |  |  ├> \\"Ccc\\"
|  |  |  |  ├> \\"\\"
|  |  |  |  └> \\"A\\"
|  |  |  |     └> \\"\\"
|  |  |  └> \\"A\\"
|  |  |     └> \\"\\"
|  |  └> \\"Bb\\"
|  |     ├> \\"\\"
|  |     └> \\"A\\"
|  |        └> \\"\\"
|  ├> \\"ABbCcc\\"
|  |  ├> \\"BbCcc\\"
|  |  |  ├> \\"Ccc\\"
|  |  |  |  ├> \\"\\"
|  |  |  |  └> \\"A\\"
|  |  |  |     └> \\"\\"
|  |  |  ├> \\"ACcc\\"
|  |  |  |  ├> \\"Ccc\\"
|  |  |  |  |  ├> \\"\\"
|  |  |  |  |  └> \\"A\\"
|  |  |  |  |     └> \\"\\"
|  |  |  |  └> \\"A\\"
|  |  |  |     └> \\"\\"
|  |  |  └> \\"Bb\\"
|  |  |     ├> \\"\\"
|  |  |     └> \\"A\\"
|  |  |        └> \\"\\"
|  |  ├> \\"A\\"
|  |  |  └> \\"\\"
|  |  ├> \\"ACcc\\"
|  |  |  ├> \\"Ccc\\"
|  |  |  |  ├> \\"\\"
|  |  |  |  └> \\"A\\"
|  |  |  |     └> \\"\\"
|  |  |  └> \\"A\\"
|  |  |     └> \\"\\"
|  |  └> \\"ABb\\"
|  |     ├> \\"Bb\\"
|  |     |  ├> \\"\\"
|  |     |  └> \\"A\\"
|  |     |     └> \\"\\"
|  |     └> \\"A\\"
|  |        └> \\"\\"
|  ├> \\"Bb\\"
|  |  ├> \\"\\"
|  |  └> \\"A\\"
|  |     └> \\"\\"
|  ├> \\"BbCcc\\"
|  |  ├> \\"Ccc\\"
|  |  |  ├> \\"\\"
|  |  |  └> \\"A\\"
|  |  |     └> \\"\\"
|  |  ├> \\"ACcc\\"
|  |  |  ├> \\"Ccc\\"
|  |  |  |  ├> \\"\\"
|  |  |  |  └> \\"A\\"
|  |  |  |     └> \\"\\"
|  |  |  └> \\"A\\"
|  |  |     └> \\"\\"
|  |  └> \\"Bb\\"
|  |     ├> \\"\\"
|  |     └> \\"A\\"
|  |        └> \\"\\"
|  └> \\"BbBb\\"
|     ├> \\"Bb\\"
|     |  ├> \\"\\"
|     |  └> \\"A\\"
|     |     └> \\"\\"
|     ├> \\"ABb\\"
|     |  ├> \\"Bb\\"
|     |  |  ├> \\"\\"
|     |  |  └> \\"A\\"
|     |  |     └> \\"\\"
|     |  └> \\"A\\"
|     |     └> \\"\\"
|     └> \\"Bb\\"
|        ├> \\"\\"
|        └> \\"A\\"
|           └> \\"\\"
├> \\"ABbBbCcc\\"
|  ├> \\"BbBbCcc\\"
|  |  ├> \\"BbCcc\\"
|  |  |  ├> \\"Ccc\\"
|  |  |  |  ├> \\"\\"
|  |  |  |  └> \\"A\\"
|  |  |  |     └> \\"\\"
|  |  |  ├> \\"ACcc\\"
|  |  |  |  ├> \\"Ccc\\"
|  |  |  |  |  ├> \\"\\"
|  |  |  |  |  └> \\"A\\"
|  |  |  |  |     └> \\"\\"
|  |  |  |  └> …
|  |  |  └> …
|  |  └> …
|  └> …
└> …"
`;
